src/codegen/compile.ml

```diff
@@ -4938,6 +4938,32 @@ module IC = struct
     compile_eq_const 0l ^^
     E.else_trap_with env "not a self-call"
 
+  let assert_caller_self_or_controller env =
+    let (set_len1, get_len1) = new_local env "len1" in
+    let (set_len2, get_len2) = new_local env "len2" in
+    let (set_str1, get_str1) = new_local env "str1" in
+    let (set_str2, get_str2) = new_local env "str2" in
+    system_call env "canister_self_size" ^^ set_len1 ^^
+    system_call env "msg_caller_size" ^^ set_len2 ^^
+    get_len2 ^^ Blob.dyn_alloc_scratch env ^^ set_str2 ^^
```

Comment:
The only risk that remains is that the GC trigger also allocates dynamic memory and could not work if heap is completely full. I thought that we could may grant the general memory reserve for the GC trigger (https://github.com/dfinity/motoko/pull/4305), but it may pose the risk that (if something would be broken with the GC trigger), a series of GC trigger calls may again eat up that reserve...

Reply:
I think I've manage to remove all allocation, by implementing the caller assertion using just stack allocation, not deserializing the trivial argument and only sending a pre-allocated response (not sure that's still pre-allocated in your orthogonal persistence world though). It does mean users can send any old junk payload. One improvement would be to check the argument is literally DIDL\x00\x00 using just stack allocation, but that wouldn't allow for subtyping so is also a hack.

I've added a test for the access control check - wasn't worth refactoring with `assert_calller_self` w/o totally obscuring the logic.

The only thing I'm not sure of is whether you still want to lift the reserve or not, now that this is heap-alloc free. If not, then should be revert the new Lifecycle state? 

Feel free to take over this PR, merge as you see fit (just drop me a message on slack if you do)

Reply:
This is great and I believe safer than using the memory reserve. Thanks a lot.
With this, I can again remove the special memory reserve handling for the GC trigger.